tests/test_websocket_bootstrap.c

```diff
@@ -0,0 +1,635 @@
+    (void)ctx;
+    ASSERT_SUCCESS(s_tester_init(allocator));
+
+    int websocket_connect_error_code;
+    ASSERT_SUCCESS(s_drive_websocket_connect(&websocket_connect_error_code));
+    ASSERT_INT_EQUALS(AWS_ERROR_SUCCESS, websocket_connect_error_code);
+
+    ASSERT_SUCCESS(s_tester_clean_up());
+    return AWS_OP_SUCCESS;
+}
+
+/* Function to be reused by all the "fail at step X" tests. */
+static int s_websocket_boot_fail_at_step_test(struct aws_allocator *alloc, void *ctx, enum boot_step fail_at_step) {
+    (void)ctx;
+    s_tester.fail_at_step = fail_at_step;
+    ASSERT_SUCCESS(s_tester_init(alloc));
+
+    int websocket_connect_error_code;
+    ASSERT_SUCCESS(s_drive_websocket_connect(&websocket_connect_error_code));
+    ASSERT_INT_EQUALS(fail_at_step, websocket_connect_error_code);
+
+    ASSERT_SUCCESS(s_tester_clean_up());
+    return AWS_OP_SUCCESS;
+}
+
+TEST_CASE(websocket_boot_fail_at_http_connect) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_HTTP_CONNECT);
+}
+
+TEST_CASE(websocket_boot_fail_at_http_connect_error) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_HTTP_CONNECT_COMPLETE);
+}
+
+TEST_CASE(websocket_boot_fail_at_new_request) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_REQUEST_NEW);
+}
+
+TEST_CASE(websocket_boot_fail_at_response_error) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_REQUEST_COMPLETE);
+}
+
+TEST_CASE(websocket_boot_fail_at_response_status) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_VALIDATE_RESPONSE_STATUS);
+}
+
+TEST_CASE(websocket_boot_fail_at_new_handler) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_WEBSOCKET_NEW);
+}
+
+TEST_CASE(websocket_boot_report_unexpected_http_shutdown) {
+    return s_websocket_boot_fail_at_step_test(allocator, ctx, BOOT_STEP_HTTP_SHUTDOWN);
+}
+
+#define TIMEBOMB_MAX_TIMER 0xFFFF
+
+struct timebomb_impl {
+    struct aws_allocator *real_allocator;
+    struct aws_atomic_var timer;
+};
+
+static void *s_timebomb_mem_acquire(struct aws_allocator *alloc, size_t size) {
+    struct timebomb_impl *timebomb = alloc->impl;
+
+    /* ALL allocations should fail after timer reaches zero.
+     * But timer is unsigned atomic.
+     * Therefore, when timer wraps around 0 and becomes huge, continue failing. */
+    size_t timer = aws_atomic_fetch_sub(&timebomb->timer, 1);
+    if (timer == 0 || timer > TIMEBOMB_MAX_TIMER) {
+        return NULL;
+    }
+
+    return timebomb->real_allocator->mem_acquire(timebomb->real_allocator, size);
+}
+
+static void s_timebomb_mem_release(struct aws_allocator *alloc, void *ptr) {
+    struct timebomb_impl *timebomb = alloc->impl;
+    timebomb->real_allocator->mem_release(timebomb->real_allocator, ptr);
+}
+
+/* Run connection process with an allocator that fakes running out of memory after N allocations. */
+TEST_CASE(websocket_boot_fail_because_oom) {
```

Comment:
This is amazing.  We should do this for.... everything.

Reply:
In fact, can we move the allocator and the test pattern (as a macro with invocation hooks maybe) up into common?

Reply:
I have a lot I need to get done this week. Can I leave that as an exercise for the *next* person who wants to try this out?

Reply:
That's fair. It's going to cause mayhem I bet when we try to use it, so let's be more intentional about that.